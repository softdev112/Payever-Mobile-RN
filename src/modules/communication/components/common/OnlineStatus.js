import { StyleSheet, View } from 'ui';

export default function OnlineStatus({ isOnline, size, style }: PropTypes) {
  const statusStyles = [styles.container, style];
  if (isOnline) {
    statusStyles.push(styles.onlineStatus);
  } else {
    statusStyles.push(styles.offlineStatus);
  }

  if (size !== undefined) {
    statusStyles.push({
      height: size,
      width: size,
      borderRadius: size / 2,
    });
  }

  return (
    <View style={statusStyles} />
  );
}

const styles = StyleSheet.create({
  container: {
    borderColor: '#d8d8d8',
    borderRadius: 4,
    borderWidth: 1,
    height: 8,
    width: 8,
  },

  offlineStatus: {
    backgroundColor: '#FFF',
    borderWidth: 1,
  },

  onlineStatus: {
    backgroundColor: '#75b636',
    borderWidth: 0,
  },
});

type PropTypes = {
  online: boolean;
};