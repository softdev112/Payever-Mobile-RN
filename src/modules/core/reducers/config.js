import { Map } from 'immutable';
import config from '../../../config';

const initialState = Map(config);

/**
 * Share config through store
 */
export default function configReducer(state = initialState) {
  return state;
}