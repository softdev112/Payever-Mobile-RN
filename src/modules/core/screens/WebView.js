import { Component } from 'react';
import { WebView as ReactWebView } from 'react-native';

export default class WebView extends Component {
  render() {
    const { url } = this.props;
    return (
      <ReactWebView source={{ uri: url }} />
    );
  }
}