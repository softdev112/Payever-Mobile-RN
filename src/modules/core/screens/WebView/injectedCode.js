/* eslint-disable */
export default function injectedCode(options: InjectOptions) {
  const optionsJson = JSON.stringify(options);
  return `(${injectedBody.toString()})(${optionsJson})`;
}

function injectedBody(options) {
  init(options);

  function init(options) {
    window.__DEV__ = options.isDev;

    if (options.isAddBusiness) {
      replaceSearchWithBackBtn(options);
      attachListenerOnSubmit(businessAddedListener);
    }

    if (options.platform === 'android') {
      patchSelectElementsForAndroid();
    }

    replaceHeaderButtons();

    if (__DEV__) {
      attachOnError();
      patchPostMessage();
    }

    getNavBarInfo();
  }

  function replaceHeaderButtons() {
    //noinspection ES6ConvertVarToLetConst
    var $btnMenu = document.querySelector('#user-nav-block .dropdown-toggle');
    if ($btnMenu) {
      //noinspection ES6ConvertVarToLetConst
      var $replace = $btnMenu.cloneNode(true);
      $replace.href = 'javascript:return void()';
      $replace.style.marginRight = 0;
      $replace.addEventListener('click', function(e) {
        e.preventDefault();
        e.stopImmediatePropagation();
        sendData({ command: 'show-menu' });
      });
      $btnMenu.parentNode.replaceChild($replace, $btnMenu);
    }
  }

  

  function replaceSearchWithBackBtn({ title }) {
    //noinspection ES6ConvertVarToLetConst
    const $rootNode = document.querySelector('.main-search');

    if ($rootNode) {
      $rootNode.style.display = 'none';
      const containerStyle = '\"padding-top: 5px; display: flex;' +
        'flex-direction: row; justify-content: space-between\"';

      //noinspection HtmlUnknownAttribute
      $rootNode.outerHTML =
        `<div style=${containerStyle}>
          <a class="back-selling-link" style="align-self: center" href="#">
            <svg class="icon icon-24">
              <use xlink:href="#icon-arrow-left-ios-24"></use>
            </svg>
          </a>
          <h4 style="">${title}</h4>
        </div>`;

      const $backA = document.querySelector('.back-selling-link');
      $backA.addEventListener('click', goBackListener);
    }
  }

  function patchSelectElementsForAndroid() {
    // In Android WebView Select elements
    // don't display well if they have css border just switch off css
    //noinspection ES6ConvertVarToLetConst
    var $selects = document.querySelectorAll('select');

    if (!$selects) return;
    for(let i = 0; i < $selects.length; ++i) {
      // select-custom form-control select-custom-ready
      // select-custom = {opacity: 0} it was only android issue
      $selects[i].className = 'form-control select-custom-ready';
    }
  }

  function patchPostMessage() {
    // To prevent error message in iOS WebView
    //noinspection ES6ConvertVarToLetConst
    var originalPostMessage = window.postMessage;
    //noinspection ES6ConvertVarToLetConst
    var patchedPostMessage = function(message, targetOrigin, transfer) {
      originalPostMessage(message, targetOrigin, transfer);
    };

    patchedPostMessage.toString = function() {
      //noinspection JSUnresolvedVariable
      return String(Object.hasOwnProperty).replace('hasOwnProperty', 'postMessage');
    };

    window.postMessage = patchedPostMessage;
  }

  function attachListenerOnSubmit(listener) {
    // Find submit button
    const $btn = document.querySelector('button[type="submit"]');
    
    if (!$btn) return;

    $btn.addEventListener('click', listener);
  }

  function attachOnError() {
    window.onerror = (errorMsg, url, lineNumber) => {
      sendData({
        errorMsg,
        url,
        lineNumber,
        command: 'error',
      })
    };
  }

  // Overridden window.postMessage can be not available when calling
  function sendData(data) {
    try {
      window.postMessage(JSON.stringify(data));
    } catch(e) {}
  }

  function getNavBarInfo() {
    let title = '';
    let titleImgUrl = '';

    const $titleImg = document.querySelector('.channel-title img');
    if ($titleImg) {
      titleImgUrl = $titleImg.getAttribute('src');
    }

    const $title = document.querySelector('.channel-title .title');
    if ($title) {
      title = $title.innerText;
    }

    // Send data to WebView
    if (title !== '' && titleImgUrl !== '') {
      sendData({
        command: 'navbar-info',
        title,
        titleImgUrl,
      });
    }
  }

  function goBackListener(e) {
    e.preventDefault();
    e.stopImmediatePropagation();
    sendData({ command: 'go-back' });
  }

  function businessAddedListener() {
    sendData({ command: 'add-business' });
  }
}

export function getLoaderHtml(url: string) {
  return (`
    <!doctype html>
    <html>
    <head>
      <title>Loading...</title>
      <meta charset="utf-8">
      <style>
        @-webkit-keyframes spin {
          0% {  -webkit-transform: rotate(0); }
          100% { -webkit-transform: rotate(360deg); }
        }
        @keyframes spin {
          0% { -webkit-transform: rotate(0); transform: rotate(0); }
          100% { -webkit-transform: rotate(360deg); transform: rotate(360deg); }
        }
        .spinner {
          position: absolute;
          top: 50%;
          left: 50%;
          margin: -32px;
          width: 64px;
          height: 64px;
          -webkit-animation: spin 1s linear infinite;
          animation: spin 1s linear infinite;
          -webkit-transform: translateZ(0);
          transform: translateZ(0);
        }
        .spinner path {
          fill: none;
          stroke: #e1e1e1;
          stroke-width: 1;
          stroke-linecap: butt;
          stroke-linejoin: miter;
          stroke-miterlimit: 10;
          stroke-opacity: 1;
        }
        .spinner .path2 { stroke: #0084ff;}
      </style>
    </head>
    <body>
      <svg 
        class="spinner"
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 64 64"
        preserveAspectRatio="xMidYMin slice"
        >
        <path 
          class="path1"
          d="M32,0.501c-17.369,0-31.5,14.131-31.5,31.5s14.131,31.5,31.5,31.5s31.5-14.131,31.5-31.5S49.369,0.501,32,0.501 z"
        />
        <path 
          class="path2"
          d="M32,0.501c-17.369,0-31.5,14.131-31.5,31.5"
        />
      </svg>
      <script>location.href = '${url}';</script>
    </body>
    </html>
  `);
}

type InjectOptions = {
  isDev: boolean;
  isAddBusiness: boolean;
  platform: string;
};