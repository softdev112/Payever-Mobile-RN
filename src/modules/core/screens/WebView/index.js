import { Component } from 'react';
import { WebView as ReactWebView, View, Linking } from 'react-native';
import { observer, inject } from 'mobx-react/native';
import type { Navigator } from 'react-native-navigation';
import { NavBar, StyleSheet } from 'ui';
import { log } from 'utils';

import injectedCode, { getLoaderHtml } from './injectedCode';
import WebViewError from './WebViewError';
import { showScreen, toggleMenu } from '../../../../common/Navigation';
import type AuthStore from '../../../../store/AuthStore';
import type UserProfilesStore from '../../../../store/UserProfilesStore';
import type { Config } from '../../../../config';

const BACK_ON_URLS = [
  { urlMask: '/home',    screen: 'dashboard.Dashboard', authRequired: true },
  { urlMask: '/private', screen: 'dashboard.Private',   authRequired: true },
  { urlMask: '/login',   screen: 'auth.Login' },
];

@inject('auth', 'userProfiles', 'config')
@observer
export default class WebView extends Component {
  static defaultProps = {
    enableExternalBrowser: false,
  };

  static navigatorStyle = {
    navBarHidden: true,
  };

  props: {
    auth: AuthStore;
    config: Config;
    enableExternalBrowser?: boolean;
    injectOptions?: Object;
    navigator: Navigator;
    referer?: string;
    url: string;
    userProfiles?: UserProfilesStore;
  };

  state: {
    errorMsg: string;
    isCustomNavBar: boolean;
    title: string;
    titleImgUrl: string;
  };

  $view: ReactWebView;
  injectedCode: string;

  constructor(props) {
    super(props);

    this.injectedCode = injectedCode({
      isDev: __DEV__,
      ...props.injectOptions,
    });

    this.state = {
      errorMsg: '',
      isCustomNavBar: false,
      title: '',
      titleImgUrl: '',
    };
  }

  onError({ nativeEvent }) {
    this.setState({
      errorMsg: nativeEvent.description,
    });
  }

  onGoBack() {
    //noinspection JSUnresolvedFunction
    this.$view.goBack();
  }

  onLoadStart({ nativeEvent }) {
    const { auth } = this.props;

    if (nativeEvent.url.startsWith('react-js-navigation')) {
      return;
    }

    this.setState({ isCustomNavBar: false });

    if (nativeEvent.url
      && !nativeEvent.url.startsWith(this.props.config.siteUrl)
      && !nativeEvent.url.includes('about:blank')
      && !nativeEvent.url.startsWith('data:text')) {
      if (this.props.enableExternalBrowser) {
        // If enableExternalBrowser === true run link in a external browser
        //noinspection JSUnresolvedFunction
        this.$view.stopLoading();
        Linking.openURL(nativeEvent.url).catch(log.error);
        return;
      }

      // Switch on custom NavBar for navigation purpose on external sites
      // if they open in WebView
      this.setState({ isCustomNavBar: true });
      return;
    }


    // Apply pat
    // Process submit events ONLY iOS UIWebKit there is no such
    // field in android WebKit nativeEvent
    if (nativeEvent.navigationType === 'formsubmit') {
      if (nativeEvent.url.endsWith('/create-business')) {
        this.refreshBusinessesWithTimeout();
      }

      // After create-business it goes to business home just block
      if (nativeEvent.url.endsWith('/home')) {
        //noinspection JSUnresolvedFunction
        this.$view.stopLoading();
        this.props.navigator.pop({ animated: true });
        return;
      }
    }

    BACK_ON_URLS.forEach((url) => {
      if (nativeEvent.url.endsWith(url.urlMask)) {
        //noinspection JSUnresolvedFunction
        this.$view.stopLoading();

        if (url.authRequired && !auth.isLoggedIn) {
          showScreen('auth.Login');
          return;
        }
        showScreen(url.screen);
      }
    });
  }

  onMessage({ nativeEvent }) {
    const siteUrl = this.props.config.siteUrl;
    if (nativeEvent.url && !nativeEvent.url.startsWith(siteUrl)) return;

    const data = nativeEvent.data;
    const object = JSON.parse(data);
    switch (object.command) {
      case 'show-menu': {
        toggleMenu(this.props.navigator);
        break;
      }
      case 'error': {
        log.warn(
          `WebView error: ${data.errorMsg} at ${data.url}:${data.lineNumber}`
        );
        this.props.navigator.pop({ animated: true });
        break;
      }

      case 'add-business':
        this.props.navigator.pop({ animated: true });
        this.refreshBusinessesWithTimeout();
        break;

      case 'go-back':
        this.props.navigator.pop({ animated: true });
        break;

      case 'navbar-info': {
        this.setState({
          title: object.title,
          titleImgUrl: siteUrl + object.titleImgUrl,
        });
        console.log(siteUrl + object.titleImgUrl);
        break;
      }

      default: {
        log.warn(`Unknown webview command ${object.command}`);
      }
    }
  }

  refreshBusinessesWithTimeout() {
    if (!this.refreshTimer) {
      this.refreshTimer = setTimeout(() => {
        // Reload businesses with some delay
        //noinspection JSIgnoredPromiseFromCall
        this.props.userProfiles.load();
      }, 1200);
    }
  }

  renderError() {
    return (
      <WebViewError
        message={this.state.errorMsg}
        navigator={this.props.navigator}
      />
    );
  }

  render() {
    const { url, referer } = this.props;
    const { errorMsg, isCustomNavBar, title, titleImgUrl } = this.state;
    const topInset = isCustomNavBar ? -75 : 0;

    let source;
    if (referer) {
      const headers = {};
      if (referer) {
        headers.Referer = referer;
      }
      source = { headers, uri: url };
    } else {
      source = { html: getLoaderHtml(url) };
    }

    const titleImgSource = titleImgUrl ? { uri: titleImgUrl } : null;

    return (
      <View style={styles.container}>
        {isCustomNavBar && !errorMsg && (
          <NavBar>
            <NavBar.Back onPress={::this.onGoBack} />
            <NavBar.Title source={titleImgSource} title={title} />
            <NavBar.Menu />
          </NavBar>
        )}
        <ReactWebView
          contentInset={{ top: topInset, left: 0, bottom: 0, right: 0 }}
          source={source}
          ref={$v => this.$view = $v}
          onError={::this.onError}
          onLoadStart={::this.onLoadStart}
          onMessage={::this.onMessage}
          javaScriptEnabled
          domStorageEnabled
          startInLoadingState={false}
          injectedJavaScript={this.injectedCode}
          renderError={::this.renderError}
          bounces={false}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    '@media ios': {
      marginTop: 15,
    },
  },
});