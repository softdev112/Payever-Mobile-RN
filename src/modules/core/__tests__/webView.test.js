/* eslint-disable max-len */
import React from 'react';
import renderer from 'react-test-renderer';
import { Linking } from 'react-native';
import { shallow } from 'enzyme';
import { cacheHelper, networkHelper, log } from 'utils';
import config from '../../../config';
import WebView from '../screens/WebView';
import Store from '../../../store';
import navigator from '../../../../__mocks__/navigator';
import { showScreen, toggleMenu } from '../../../common/Navigation';

jest.mock('../../../common/utils/cacheHelper')
  .mock('../../../common/utils/networkHelper')
  .mock('react-native-logging')
  .mock('../../../common/Navigation')
  .mock('Linking', () => ({
    openURL: jest.fn(() => Promise.resolve()),
  }));

describe('modules/core/screens/WebView', () => {
  let store;
  let profiles;
  let auth;

  beforeAll(() => {
    cacheHelper.isCacheUpToDate.mockImplementation(() => true);
    networkHelper.isConnected.mockImplementation(() => true);
  });

  beforeEach(() => {
    store = new Store(config);
    profiles = store.profiles;
    auth = store.auth;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('WebView should render correctly if network connection available and should NOT call navigator.showModal', () => {
    const tree = renderer.create(
      <WebView
        profiles={profiles}
        auth={auth}
        config={store.config}
        navigator={navigator}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();

    // componentWillMount
    expect(networkHelper.isConnected).toHaveBeenCalled();
    expect(navigator.showModal).not.toHaveBeenCalled();
  });

  it('WebView/componentWillMount() should call navigator.showModal if network connection NOT available', async () => {
    jest.useFakeTimers();
    networkHelper.isConnected.mockImplementationOnce(() => false);

    await renderer.create(
      <WebView
        profiles={profiles}
        auth={auth}
        config={store.config}
        navigator={navigator}
      />
    ).toJSON();

    expect(networkHelper.isConnected).toHaveBeenCalled();
    expect(navigator.showModal).toHaveBeenCalled();
    expect(navigator.showModal).toHaveBeenCalledWith({
      passProps: {
        message: networkHelper.errorConnection,
        onBack: expect.any(Function),
      },
      screen: 'core.ErrorPage',
    });
    expect(setTimeout).toHaveBeenCalledTimes(1);

    jest.clearAllTimers();
  });

  describe('WebView/onLoadStart({ nativeEvent })', () => {
    it('WebView/onLoadStart({ nativeEvent }) should call Linking.openURL if it comes from external url', () => {
      const openUrlSpy = jest.spyOn(Linking, 'openURL');

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: 'https://www.github.com',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(openUrlSpy).toHaveBeenCalled();
      expect(openUrlSpy).toHaveBeenCalledWith('https://www.github.com');
    });

    it('WebView/onLoadStart({ nativeEvent }) should NOT call Linking.openURL if it comes from payever url and user logged In', () => {
      const openUrlSpy = jest.spyOn(Linking, 'openURL');

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/some_url',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(openUrlSpy).not.toHaveBeenCalled();
    });

    it('WebView/onLoadStart({ nativeEvent }) should call navigator.pop if it go to url ends with /home and user loggedIn', () => {
      auth.isLoggedIn = true;

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/home',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(navigator.pop).toHaveBeenCalled();
      expect(navigator.pop).toHaveBeenCalledWith({ animated: true });
    });

    it('WebView/onLoadStart({ nativeEvent }) should NOT call navigator.pop and should call showScreen if url ends with /home and user NOT loggedIn', () => {
      auth.isLoggedIn = false;

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/home',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(navigator.pop).not.toHaveBeenCalled();
      expect(showScreen).toHaveBeenCalled();
      expect(showScreen).toHaveBeenCalledWith('auth.Login');
    });

    it('WebView/onLoadStart({ nativeEvent }) should call navigator.pop if it go to url ends with /private and user loggedIn', () => {
      auth.isLoggedIn = true;

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/private',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(navigator.pop).toHaveBeenCalled();
      expect(navigator.pop).toHaveBeenCalledWith({ animated: true });
    });

    it('WebView/onLoadStart({ nativeEvent }) should NOT call navigator.pop and should call showScreen if url ends with /private and user NOT loggedIn', () => {
      auth.isLoggedIn = false;

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/private',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(navigator.pop).not.toHaveBeenCalled();
      expect(showScreen).toHaveBeenCalled();
      expect(showScreen).toHaveBeenCalledWith('auth.Login');
    });

    it('WebView/onLoadStart({ nativeEvent }) should call navigator.pop if it go to url ends with /login and user loggedIn', () => {
      auth.isLoggedIn = true;

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/login',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(navigator.pop).toHaveBeenCalled();
      expect(navigator.pop).toHaveBeenCalledWith({ animated: true });
    });

    it('WebView/onLoadStart({ nativeEvent }) should call navigator.pop and should NOT call showScreen if url ends with /login and user NOT loggedIn', () => {
      auth.isLoggedIn = false;

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/login',
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onLoadStart(event);

      expect(navigator.pop).toHaveBeenCalled();
      expect(navigator.pop).toHaveBeenCalledWith({ animated: true });
      expect(showScreen).not.toHaveBeenCalled();
    });
  });

  describe('WebView/onMessage({ nativeEvent })', () => {
    it('WebView/onMessage({ nativeEvent }) should call parse command if it gets bad url or url not from payever', () => {
      const jsonSpy = jest.spyOn(JSON, 'parse');

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event1 = {
        nativeEvent: {
          url: 'https://www.github.com',
        },
      };

      const event2 = {
        nativeEvent: {
          url: '',
        },
      };

      const event3 = {
        nativeEvent: {},
      };

      const event4 = {
        nativeEvent: {
          url: config.siteUrl + '/about:blank',
        },
      };

      const webView = wrapper.dive().find('WebView');

      webView.props()
        .onMessage(event1);
      webView.props()
        .onMessage(event2);
      webView.props()
        .onMessage(event3);
      webView.props()
        .onMessage(event4);

      expect(jsonSpy).not.toHaveBeenCalled();
    });

    it('WebView/onMessage({ nativeEvent }) should call toggleMenu if get command show-menu', () => {
      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/common',
          data: JSON.stringify({ command: 'show-menu' }),
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onMessage(event);

      expect(toggleMenu).toHaveBeenCalled();
      expect(toggleMenu).toHaveBeenCalledWith(navigator);
    });

    it('WebView/onMessage({ nativeEvent }) should call navigator.pop if get command error', () => {
      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/common',
          data: JSON.stringify({ command: 'error' }),
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onMessage(event);

      expect(navigator.pop).toHaveBeenCalled();
      expect(navigator.pop).toHaveBeenCalledWith({ animated: true });
    });

    it('WebView/onMessage({ nativeEvent }) should call log.warn if get unknown command', () => {
      log.warn = jest.fn();

      const wrapper = shallow(
        <WebView
          profiles={profiles}
          auth={auth}
          config={store.config}
          navigator={navigator}
          enableExternalBrowser
        />
      );

      const event = {
        nativeEvent: {
          url: config.siteUrl + '/common',
          data: JSON.stringify({ command: 'unknown' }),
        },
      };

      wrapper.dive()
        .find('WebView')
        .props()
        .onMessage(event);

      expect(log.warn).toHaveBeenCalled();
      expect(log.warn).toHaveBeenCalledWith('Unknown webview command unknown');
    });
  });
});