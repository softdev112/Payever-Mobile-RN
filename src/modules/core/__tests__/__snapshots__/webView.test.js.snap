// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/core/screens/WebView WebView should render correctly if network connection available and should NOT call navigator.showModal 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 15,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTWebView
      allowsInlineMediaPlayback={undefined}
      automaticallyAdjustContentInsets={undefined}
      bounces={false}
      contentInset={
        Object {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        }
      }
      dataDetectorTypes={undefined}
      decelerationRate={undefined}
      injectedJavaScript="(function injectedBody(options){init(options);function init(options){window.__DEV__=options.isDev;if(options.platform==='android'){patchSelectElementsForAndroid();}replaceHeaderButtons();if(__DEV__){attachOnError();patchPostMessage();}getNavBarInfo();patchAddApp();setTimeout(function(){return sendData({command:'hide-loader'},500);});}function replaceHeaderButtons(){var $btnMenu=document.querySelector('#user-nav-block .dropdown-toggle');if($btnMenu){var $replace=$btnMenu.cloneNode(true);$replace.href='javascript:return void()';$replace.style.marginRight=0;$replace.addEventListener('click',function(e){e.preventDefault();e.stopImmediatePropagation();sendData({command:'show-menu'});});$btnMenu.parentNode.replaceChild($replace,$btnMenu);}var $backBtn=document.querySelector('.btn-primary.btn-link');if($backBtn&&$backBtn.href&&$backBtn.href.includes('#')){$replace=$backBtn.cloneNode(true);$replace.href=$backBtn.href.split('#')[0];$replace.style.marginRight=0;$backBtn.parentNode.replaceChild($replace,$backBtn);}}function patchAddApp(){function searchModal(){var $dialogFooter=document.querySelector('.edit-campaign-add-item > .modal-content > .modal-footer');if($dialogFooter){var $submitButton=$dialogFooter.querySelector('.btn-link.submits-modal');function submitHandler(){if($submitButton){$submitButton.click();}}var $topFooter=$dialogFooter.cloneNode(true);$topFooter.style.position='fixed';$topFooter.style.right=0;$topFooter.style.top=0;$topFooter.style.borderBottom='1px solid #e1e1e1';$topFooter.style.backgroundColor='white';$topFooter.style.width='100%';$topFooter.style.padding='8px';$topFooter.style.paddingRight='0px';var $topSubmitBtn=$topFooter.querySelector('.btn-link.submits-modal');if($topSubmitBtn){$topSubmitBtn.addEventListener('click',submitHandler);}var $dialog=document.querySelector('.edit-campaign-add-item');$dialog.appendChild($topFooter);}setTimeout(searchModal,2000);}setTimeout(searchModal,2000);}function patchSelectElementsForAndroid(){var $selects=document.querySelectorAll('select');if(!$selects)return;for(var i=0;i<$selects.length;++i){$selects[i].className='form-control select-custom-ready';}}function patchPostMessage(){var originalPostMessage=window.postMessage;var patchedPostMessage=function patchedPostMessage(message,targetOrigin,transfer){originalPostMessage(message,targetOrigin,transfer);};patchedPostMessage.toString=function(){return String(Object.hasOwnProperty).replace('hasOwnProperty','postMessage');};window.postMessage=patchedPostMessage;}function attachOnError(){window.onerror=function(errorMsg,url,lineNumber){sendData({errorMsg:errorMsg,url:url,lineNumber:lineNumber,command:'error'});};}function sendData(data){try{window.postMessage(JSON.stringify(data));}catch(e){}}function getNavBarInfo(){var title='';var titleImgUrl='';var $titleImg=document.querySelector('.channel-title img');if($titleImg){titleImgUrl=$titleImg.getAttribute('src');}var $title=document.querySelector('.channel-title .title');if($title){title=$title.innerText;}if(title!==''&&titleImgUrl!==''){sendData({title:title,titleImgUrl:titleImgUrl,command:'navbar-info'});}}})({\\"isDev\\":true})"
      mediaPlaybackRequiresUserAction={undefined}
      messagingEnabled={true}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onShouldStartLoadWithRequest={undefined}
      scalesPageToFit={undefined}
      scrollEnabled={undefined}
      source={
        Object {
          "html": "
            <!doctype html>
            <html>
            <head>
              <title>Loading...</title>
              <meta charset=\\"utf-8\\">
              <style>
                @-webkit-keyframes spin {
                  0% {  -webkit-transform: rotate(0); }
                  100% { -webkit-transform: rotate(360deg); }
                }
                @keyframes spin {
                  0% { -webkit-transform: rotate(0); transform: rotate(0); }
                  100% { -webkit-transform: rotate(360deg); transform: rotate(360deg); }
                }
                .spinner {
                  position: absolute;
                  top: 50%;
                  left: 50%;
                  margin: -32px;
                  width: 64px;
                  height: 64px;
                  -webkit-animation: spin 1s linear infinite;
                  animation: spin 1s linear infinite;
                  -webkit-transform: translateZ(0);
                  transform: translateZ(0);
                }
                .spinner path {
                  fill: none;
                  stroke: #e1e1e1;
                  stroke-width: 1;
                  stroke-linecap: butt;
                  stroke-linejoin: miter;
                  stroke-miterlimit: 10;
                  stroke-opacity: 1;
                }
                .spinner .path2 { stroke: #0084ff;}
              </style>
            </head>
            <body>
              <svg 
                class=\\"spinner\\"
                xmlns=\\"http://www.w3.org/2000/svg\\"
                viewBox=\\"0 0 64 64\\"
                preserveAspectRatio=\\"xMidYMin slice\\"
                >
                <path 
                  class=\\"path1\\"
                  d=\\"M32,0.501c-17.369,0-31.5,14.131-31.5,31.5s14.131,31.5,31.5,31.5s31.5-14.131,31.5-31.5S49.369,0.501,32,0.501 z\\"
                />
                <path 
                  class=\\"path2\\"
                  d=\\"M32,0.501c-17.369,0-31.5,14.131-31.5,31.5\\"
                />
              </svg>
              <script>location.href = 'undefined';</script>
            </body>
            </html>
          ",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 4,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "paddingLeft": 1,
            "paddingTop": 1,
            "transform": Array [
              Object {
                "rotate": "0deg",
              },
            ],
          }
        }
      >
        <View
          disabled={false}
          layers={
            Array [
              Object {
                "component": "vector",
                "source": "spinner-container",
                "style": Object {
                  "color": "#e1e1e1",
                },
              },
              Object {
                "component": "vector",
                "source": "spinner-roller",
                "style": Object {
                  "color": "#0084ff",
                },
              },
            ]
          }
          source="spinner"
          style={
            Array [
              Object {
                "height": 65,
                "width": 65,
              },
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            source="spinner-container"
            style={
              Array [
                Object {
                  "fontFamily": "payeverIcons",
                  "fontSize": 64,
                },
                Array [
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "left": 0,
                      "position": "absolute",
                      "top": 0,
                    },
                    Object {
                      "color": "#e1e1e1",
                    },
                  ],
                  undefined,
                ],
              ]
            }
          >
            
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            source="spinner-roller"
            style={
              Array [
                Object {
                  "fontFamily": "payeverIcons",
                  "fontSize": 64,
                },
                Array [
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "left": 0,
                      "position": "absolute",
                      "top": 0,
                    },
                    Object {
                      "color": "#0084ff",
                    },
                  ],
                  undefined,
                ],
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
